pkg/controllers/sync/controller.go

```diff
@@ -190,133 +153,132 @@ func newSyncController(
-
-	s.worker = worker.NewReconcileWorker(
+	s.worker = worker.NewReconcileWorker[common.QualifiedName](
+		SyncControllerName,
+		nil,
 		s.reconcile,
 		worker.RateLimiterOptions{},
 		controllerConfig.WorkerCount,
 		controllerConfig.Metrics,
-		deliverutil.NewMetricTags("sync-worker", typeConfig.GetTargetType().Kind),
 	)
 
-	// TODO: do we need both clusterWorker and clusterQueue?
-	s.clusterWorker = worker.NewReconcileWorker(s.reconcileCluster, worker.RateLimiterOptions{}, 1, controllerConfig.Metrics,
-		deliverutil.NewMetricTags("sync-cluster-worker", typeConfig.GetTargetType().Kind))
+	s.clusterCascadingDeletionWorker = worker.NewReconcileWorker[common.QualifiedName](
+		SyncControllerName+"-cluster-cascading-deletion-worker",
+		nil,
+		s.reconcileClusterForCascadingDeletion,
+		worker.RateLimiterOptions{},
+		1,
+		controllerConfig.Metrics,
+	)
 
 	// Build queue for triggering cluster reconciliations.
-	s.clusterQueue = workqueue.NewNamedDelayingQueue("sync-controller-cluster-queue")
-
-	targetAPIResource := typeConfig.GetTargetType()
-
-	// Federated informer for resources in member clusters
-	s.informer, err = util.NewFederatedInformer(
-		controllerConfig,
-		client,
-		configCopy,
-		&targetAPIResource,
-		func(obj pkgruntime.Object) {
-			qualifiedName := common.NewQualifiedName(obj)
-			s.worker.EnqueueWithDelay(qualifiedName, s.memberObjectEnqueueDelay)
+	s.clusterReadinessTransitionQueue = workqueue.NewNamedDelayingQueue("sync-controller-cluster-readiness-transition-queue")
+
+	if err := s.fedInformerManager.AddEventHandlerGenerator(&informermanager.EventHandlerGenerator{
+		Predicate: informermanager.RegisterOncePredicate,
+		Generator: func(ftc *fedcorev1a1.FederatedTypeConfig) cache.ResourceEventHandler {
+			return util.NewTriggerOnAllChanges(func(o pkgruntime.Object) {
+				obj := o.(*unstructured.Unstructured)
+
+				ftc, exists := s.ftcManager.GetResourceFTC(obj.GroupVersionKind())
+				if !exists {
+					return
+				}
+
+				federatedName := common.QualifiedName{
+					Namespace: obj.GetNamespace(),
+					Name:      util.GenerateFederatedObjectName(obj.GetName(), ftc.GetName()),
+				}
+				s.worker.EnqueueWithDelay(federatedName, s.memberObjectEnqueueDelay)
+			})
 		},
-		&util.ClusterLifecycleHandlerFuncs{
-			ClusterAvailable: func(cluster *fedcorev1a1.FederatedCluster) {
-				// When new cluster becomes available process all the target resources again.
-				s.clusterWorker.EnqueueObject(cluster)
-				s.clusterQueue.AddAfter(struct{}{}, s.clusterAvailableDelay)
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add event handler generator: %w", err)
+	}
+
+	if err := s.fedInformerManager.AddClusterEventHandlers(
+		&informermanager.ClusterEventHandler{
+			Predicate: func(oldCluster, newCluster *fedcorev1a1.FederatedCluster) bool {
+				// Enqueue cluster when it's marked for deletion to ensure cascading deletion
+				return oldCluster != nil && newCluster != nil &&
```

Comment:
Do we need to consider the case where a cluster was marked for deletion when the controller-manager was not running? In this case `oldCluster` will be nil and `newCluster.GetDeletionTimestamp().IsZero()` will be true.